File: lintcode/python/1474_shortest_phrase.py
#!/usr/bin/python -t

# two pointers and pre_sum


class Solution:
    """
    @param k: The number of words in the article
    @param lim: TThe minimum number of words a phrase should contain 
    @param str: The article
    @return: Return the length of shortest phrase
    """
    def getLength(self, k, lim, str):
        # Write your code here
        n = len(str)
        ret = 1e9
        
        len_list = [0] * n
        
        for i in range(n):
            len_list[i] = len(str[i])
                
        left = 0
        val = 0
        
        for right in range(n):
            val += len_list[right]
            
            while right-left >= k and val - len_list[left] >= lim:
                val -= len_list[left]
                left += 1
                
            if right-left+1>= k and val >= lim:
                ret = min(ret, val)
            
        return ret
        

File: lintcode/golang/1474_shortest_phrase.go
/**
 * @param k: The number of words in the article
 * @param lim: TThe minimum number of words a phrase should contain 
 * @param str: The article
 * @return: Return the length of shortest phrase
 */

const MaxUint = ^uint(0) 
const MinUint = 0 
const MaxInt = int(MaxUint >> 1) 

func getLength (k int, lim int, str []string) int {
    // Write your code here
    s := make([]int, len(str))
    ret := MaxInt
    
    for i, w := range str {
        s[i] = len(w)
    }
    
    l := 0
    sum := 0
    
    for r := 0; r < len(s); r++ {
        sum += s[r]
        
        for r - l >= k && sum - s[l] >= lim {
            sum -= s[l]
            l++
        }
        
        if r - l + 1 >= k && sum >= lim {
            ret = min(ret, sum)
        }
        
    }
    
    return ret
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
